x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/ShardMultiSearchAction.java

```diff
@@ -0,0 +1,244 @@
+        protected ShardsIterator shards(ClusterState state, InternalRequest request) {
+            String index = request.concreteIndex();
+            IndexRoutingTable indexRouting = state.routingTable().index(index);
+            int numShards = indexRouting.shards().size();
+            if (numShards != 1) {
+                throw new IllegalStateException("index [" + index + "] should have 1 shard, but has " + numShards + " shards");
+            }
+            return indexRouting.shard(0).shardsRandomIt();
+        }
+
+        @Override
+        protected MultiSearchResponse shardOperation(Request request, ShardId shardId) throws IOException {
+            final long nowInMillis = System.currentTimeMillis();
+            final IndexService indexService = indicesService.indexService(shardId.getIndex());
+            final IndexShard indexShard = indicesService.getShardOrNull(shardId);
+            try (Engine.Searcher searcher = indexShard.acquireSearcher("enrich_msearch")) {
+                final FieldsVisitor visitor = new FieldsVisitor(true);
+                final QueryShardContext context =
+                    indexService.newQueryShardContext(shardId.id(), searcher.reader(), () -> nowInMillis, null);
+                final MapperService mapperService = context.getMapperService();
+                final Text typeText = mapperService.documentMapper().typeText();
+
+                final MultiSearchResponse.Item[] items = new MultiSearchResponse.Item[request.multiSearchRequest.requests().size()];
+                for (int i = 0; i < request.multiSearchRequest.requests().size(); i++) {
+                    final SearchSourceBuilder searchSourceBuilder = request.multiSearchRequest.requests().get(i).source();
+
+                    final QueryBuilder queryBuilder = searchSourceBuilder.query();
+                    final int from = searchSourceBuilder.from() == -1 ? 0 : searchSourceBuilder.from();
+                    final int size = searchSourceBuilder.size() == -1 ? 10 : searchSourceBuilder.size();
+                    final FetchSourceContext fetchSourceContext = searchSourceBuilder.fetchSource();
+
+                    // validate that only a from, size, query and source filtering has been provided (other features are not supported):
+                    // (first unset, what is supported and then see if there is anything left)
+                    searchSourceBuilder.query(null);
+                    searchSourceBuilder.from(0);
+                    searchSourceBuilder.size(10);
+                    searchSourceBuilder.fetchSource(null);
+                    if (EMPTY_SOURCE.equals(searchSourceBuilder) == false) {
+                        throw new IllegalStateException("search request [" + Strings.toString(searchSourceBuilder) +
+                            "] is using features that is not supported");
+                    }
+
+                    final Query luceneQuery = queryBuilder.rewrite(context).toQuery(context);
```

Comment:
Is it possible for us to always provide a luceneQuery and skip this re-write step ? (would that provide any value in terms of performance ?)

Reply:
No, a lucene query can't be serialized.